ports/nrf/mphalport.c

```diff
@@ -35,6 +35,76 @@
 #include "nrfx_errors.h"
 #include "nrfx_config.h"
 
+#if MICROPY_PY_TIME_USE_RTC_BASE
+#include "nrfx_rtc.h"
+#endif
+
+#if MICROPY_PY_TIME_USE_TICKER_BASE
+#include "ticker.h"
+#endif
+
+
+#if MICROPY_PY_TIME_TICKS
+
+#if MICROPY_PY_TIME_USE_RTC_BASE
+nrfx_rtc_t rtc1 = NRFX_RTC_INSTANCE(1);
+
+volatile mp_uint_t rtc_overflows = 0;
+
+const nrfx_rtc_config_t rtc_config_time_msec = {
+    .prescaler    = 0,
+    .reliable     = 0,
+    .tick_latency = 0,
+    #ifdef NRF51
+    .interrupt_priority = 3,
```

Comment:
Could this be to set priority 1 for nrf51 and priority 3 for others? 

I'm not having a very good example, but imagine someone using machine.RTC/machine.Timer, to issue a timed operation like poll(timeout) on regular intervals. As the callback is in IRQ priority 3, and RTC1 is in priority 3, these will chain, and the overflow event might not get through (given that mp_hal_ticks_ms() is the source of the timed operation). 

The priorities proposed should be fine with SoftDevices, as these are free for application use. And the fact that the overflow routine is this small, it should not impact any performance of the stacks.

Reply:
> Could this be to set priority 1 for nrf51 and priority 3 for others?

I have no strong opinion on this. If you think that's better I have no problem adapting it.

> As the callback is in IRQ priority 3, and RTC1 is in priority 3, these will chain, and the overflow event might not get through (given that mp_hal_ticks_ms() is the source of the timed operation).

I think you are relating to the time slot where the overflow has occured but something prevents the rtc irq handler being called (e.g. by a already running irq)? During this time,  `mp_hal_ticks_ms()` would return an invalid value since the counter's value is 0 and the overflow counter hasn't increased. Assuming that `mp_hal_ticks_ms()` is not called in a irq context then it's not a problem but otherwise it could create a problem (very rare though, every 512 sec for a very short time and assuming that `mp_hal_ticks_ms()` is called in irq). 

Otherwise, `mp_hal_ticks_ms()` could be guarded by a check to the overflow flag... Perhaps that's the simplest solution to this.

Just wondering if nested irq's enabled?

Reply:
I agree, its not a problem if `mp_hal_ticks_ms` is not called from IRQ context. And this is a very very tiny time window to hit it, but it might lead to a deadlock.

> Just wondering if nested irq's enabled?

IRQ's for the peripherals are nested. The nrfx drivers use the `nrfx_glue.h` in the nrf-port to map the NVIC functions.

> Otherwise, mp_hal_ticks_ms() could be guarded by a check to the overflow flag... Perhaps that's the simplest solution to this. 

I like the idea :) it would make my concern about this "IRQ in IRQ" go away. Letting `mp_hal_ticks_ms()` read and clear the OVERFLOW event flag (bypassing the nrfx driver enabling the interrupt, and manually enable the event). But, how would we enforce `mp_hal_ticks_ms()` to be run at least once between overflows? IMO keeping the IRQ handler to detect overflow as you have it already is better after all, and raise the IRQ priority.

Reply:
> Letting mp_hal_ticks_ms() read and clear the OVERFLOW event flag

Yeah I was just digging through this and to make things simple: why not, on OVERFLOW detected,  just return the last valid tick value and make the irq clearing the event bit? 

We would need to store the last valid value though.

> But, how would we enforce mp_hal_ticks_ms() to be run at least once between overflows?

Ah we may have misunderstood each other. My idea was to leave the irq handler as is and make the guarding only to check if an overflow is pending. So there is no need to enforce a call.

Reply:
> IRQ's for the peripherals are nested.

hmm that makes it worse.. then even `rtc_irq_time`can be interrupted (possibly even in the nrfx HAL code part..) if that happens and the higher irq handler uses the tick method, this would lead to the same thing as before. 

I mean we could agree on not using the `mp_hal_ticks_ms()` within irqs?

Reply:
> then even rtc_irq_timecan be interrupted (possibly even in the nrfx HAL code part..)

Do you mean the other peripherals? or the RTC?

> I mean we could agree on not using the mp_hal_ticks_ms() within irqs?

I know it is not the best code pattern, but it will be code that does it. It might be hard to not do it when `machine.Timer` and `machine.RTCounter` has callbacks to fill. These are triggered from IRQ context in its given priority.

Or, maybe agree that we only call `mp_hal_ticks_ms()` from a lower IRQ or main thread? :)
This is the case for most of the machine modules. They are configured to priority 3 and 6.

Reply:
For `mp_hal_ticks_us()` (note us not ms) stm32 handles overflow both in the SysTick IRQ and in the call to `mp_hal_ticks_us()`.  And SysTick IRQ runs at the highest priority.

> Or, maybe agree that we only call `mp_hal_ticks_ms()` from a lower IRQ or main thread?

That might be a good solution.

Reply:
> Do you mean the other peripherals? or the RTC?

I meant that the RTC's irq entry point is in the nrfx lib and is eventually dispatched to the subscribed handler(s). So `rtc_irq_time` is really just the handler part. The majority of the irq time is probably spend in the nrfx part, checking and calling the handlers etc. And during this time, the irq can be preempted also.

> Or, maybe agree that we only call mp_hal_ticks_ms() from a lower IRQ or main thread? :)
This is the case for most of the machine modules. They are configured to priority 3 and 6.

I also think that's a good proposal.

Reply:
>  Or, maybe agree that we only call mp_hal_ticks_ms() from a lower IRQ or main thread? :)

sigh sorry for coming back to the subject, but I think this doesn't work either since `mp_hal_ticks_ms` can be preempted by the RTC irq. Specifically the line `return (rtc_overflows << 9) * 1000 + ((mp_uint_t)rtc1.p_reg->COUNTER * 125 / 4096);` can interrupted by the RTC irq. If that happens, the RTC counter is at 0 but the rtc_overflows counter had the old value so the caluclated value is actually wrong. It's a tiny time window, but still possible to hit. 

What could be done is to guard the calculation against the irq of the RTC like: 

- disable RTC irq. 
- make the calculation. 
- reenable RTC irq again. 
- return tick value.

Reply:
Ok I've pushed a version that does the above guarding. Let me know if the current version is so far ok. Then I'll cleanup the commit mess ;)

Reply:
see below comments for a better solution